fix: ibeacon crashing on android 8.x

dopry opened the pull request:
I updated my Nexus 6P to android 8.x, An Ionic 4 app I'm developing started crashing every time I tried to use ibeacon.  Using ADB Logcat I noticed the error...

```
FATAL EXCEPTION: AsyncTask #4
Process: com.sprygroup.automata, PID: 27596
java.lang.RuntimeException: An error occurred while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:353)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
       at java.util.concurrent.FutureTask.run(FutureTask.java:271)
       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/support/v4/content/LocalBroadcastManager;
       at org.altbeacon.beacon.BeaconLocalBroadcastProcessor.unregister(BeaconLocalBroadcastProcessor.java:82)
       at org.altbeacon.beacon.BeaconLocalBroadcastProcessor.register(BeaconLocalBroadcastProcessor.java:74)
       at org.altbeacon.beacon.service.ScanJobScheduler.ensureNotificationProcessorSetup(ScanJobScheduler.java:69)
       at org.altbeacon.beacon.service.ScanJobScheduler.schedule(ScanJobScheduler.java:122)
       at org.altbeacon.beacon.service.ScanJobScheduler.applySettingsToScheduledJob(ScanJobScheduler.java:85)
       at org.altbeacon.beacon.service.ScanJobScheduler.applySettingsToScheduledJob(ScanJobScheduler.java:92)
       at org.altbeacon.beacon.BeaconManager.applyChangesToServices(BeaconManager.java:962)
       at org.altbeacon.beacon.BeaconManager.startMonitoringBeaconsInRegion(BeaconManager.java:902)
       at com.unarin.cordova.beacon.LocationManager$16.run(LocationManager.java:832)
       at com.unarin.cordova.beacon.LocationManager$33.doInBackground(LocationManager.java:1433)
       at com.unarin.cordova.beacon.LocationManager$33.doInBackground(LocationManager.java:1427)
       at android.os.AsyncTask$2.call(AsyncTask.java:333)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       ... 4 more
Caused by: java.lang.ClassNotFoundException: Didn't find class "android.support.v4.content.LocalBroadcastManager" on path: DexPathList[[zip file "/data/app/com.sprygroup.automata-wSTYa-9sJHZxG2fs4myqrA==/base.apk"],nativeLibraryDirectories=[/data/app/com.sprygroup.automata-wSTYa-9sJHZxG2fs4myqrA==/lib/arm64, /system/lib64, /vendor/lib64]]
       at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:125)
       at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
```

I added an explicit reference to the library in the plugin.xml and it seems to have resolved the crashing issue. 

possilbly applies to: 
* https://github.com/petermetz/cordova-plugin-ibeacon/issues/335
* https://github.com/petermetz/cordova-plugin-ibeacon/issues/340
* https://github.com/petermetz/cordova-plugin-ibeacon/issues/346

dopry committed:
fix: add reference to com.android.support.support-core-utils

* including only core-utils to streamline size, 1.23MB smaller than support-v4
   in demo:  v4 size:  12.10MB, core-utils size: 10.87Mb

dopry commented:
updated to only include support-core-utils, it slimmed down the final code in the app I'm testing by 1.23MB. You can test builds with support-v4 instead to see the difference in size in your own test environment.

dopry commented:
If you want to test and you're not comfortable working with cordova plugins locally, you should be able to test by installing from my fork.

```
cordova plugin add https://github.com/dopry/cordova-plugin-ibeacon#5f4bebbfcfdf563f3d8666a57162e62a588c608f
```

jakeleveroni commented:
Thanks for doing this! I just pulled your fix into my local plugin and it fixed the crash but now I get an error:

` Native layer failed to begin monitoring: No such service ComponentInfo{io.ionic.starter/org.altbeacon.beacon.service.ScanJob}`

coming from this snippet: 
```typescript
    this.iBeacon.startMonitoringForRegion(this.beaconRegion)
      .then( _ => console.log('Native layer recieved the request to monitoring'),
             error => console.error('Native layer failed to begin monitoring: ', error) 
      );
```

Have you run into anything like this @dopry?

dopry commented:
I think that means Java couldn't find the ScanJob class in the altbeacon jar.  Is your local plugin the latest master from petermetz?

Try a platform rm, plugin rm, platform add, plugin add, then build, if you haven't already.

Then maybe try pulling my other PR to update altbeacon.

jakeleveroni commented:
@dopry sorry I got pulled over to the iOS side of things, ill try to give the other PR tonight or tomorrow. Thanks for your reply!

dopry commented:
You'll need to pull both to get a working build probably, but I'd
definitely try the platform/plugin add remove. I feel like it doesn't
always clean and rebuild properly in my local development.

On Thu, Apr 19, 2018, 4:18 PM Jacob <notifications@github.com> wrote:

> @dopry <https://github.com/dopry> sorry I got pulled over to the iOS side
> of things, ill try to give the other PR tonight or tomorrow. Thanks for
> your reply!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/petermetz/cordova-plugin-ibeacon/pull/363#issuecomment-382867363>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqOA_hOYMtMOHrdBp14UBye4O7QOrJks5tqPD1gaJpZM4TURj3>
> .
>

muhammadsamir11 commented:
I still have this issue
Native layer failed to begin monitoring: No such service ComponentInfo{io.ionic.starter/org.altbeacon.beacon.service.ScanJob}

dopry commented:
@muhammadsamir11 @jakeleveroni, I've been able to reproduce. 

I noticed in my config.xml
```
<config-file parent="/manifest/application" target="AndroidManifest.xml">
            <service android:name="org.altbeacon.beacon.service.ScanJob" android:permission="android.permission.BIND_JOB_SERVICE" />
</config-file>
<config-file parent="/manifest" target="AndroidManifest.xml">
            <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
</config-file>
```
When I remove this I get the error. I'm working on an update this will resolve this. I should have it pushed shortly. I'll update you when I'm done testing locally.

dopry committed:
fix: add new permissions required for android 8

In testing errors were thrown without these permissions.

* BIND_JOB_SERVICE

```
Unhandled Promise rejection: No such service
ComponentInfo{com.sprygroup.automata/org.altbeacon.beacon.service.ScanJob} ; Zone: <root> ; Task: ; Value:
No such service ComponentInfo{com.sprygroup.automata/org.altbeacon.beacon.service.ScanJob}
```

* RECEIVE_BOOT_COMPLETED

```
Unhandled Promise rejection: Error: requested job be persisted without holding
RECEIVE_BOOT_COMPLETED permission. ; Zone: <root> ; Task: ; Value: Error: requested job be persisted without
holding RECEIVE_BOOT_COMPLETED permission.
```

dopry commented:
@muhammadsamir11, @jakeleveroni, can you guys give the most recent push a test? It should work reasonably well for beacons while the app is in the foreground.

jakeleveroni commented:
@dopry sorry about the slow response but i just had a chance to test the latest [commit](https://github.com/petermetz/cordova-plugin-ibeacon/pull/363/commits/80e2f4bb5bff60ec4c127a5c0eefb0be56e9821f]) and it worked fine for me. Im on an Google pixel running android 8.1.0

Thanks for all your hard work!

maske94 commented:
Hi, 

I'm experiencing the same problem with Android 8. 

Since I have read that @dopry  fork solution seems to work, is there any chance to merge his changes directly into @petermetz project ? 

Thank you very much for your hard work.

dopry commented:
@petermetz, are there any issues with this PR you need me to address to get it merged. It seems like people are waiting on this functionality and I'd like to get back to pulling code from npm instead of from my git repo.

dopry commented:
remerged as https://github.com/petermetz/cordova-plugin-ibeacon/commit/928f9f47110cee40f2abb86e369c6e1fdcfa5bf6

angel-menendez commented:
@dopry  Testing in a Pixel with Android 8.1 my app is being killed in background. I was searching the error in AltBeacon and found this [https://github.com/AltBeacon/android-beacon-library/pull/674/files/a906104acc65b42bfc5b86203ae1c3e21a159636](url).

Wondering if it is a confirmation of it if not posible to monitoring beacons in background for Android 8+? Would appreciate some clarification.

Here the error captured by android 8.1.

java.lang.IllegalStateException: 
  at android.app.ContextImpl.startServiceCommon (ContextImpl.java:1521)
  at android.app.ContextImpl.startService (ContextImpl.java:1477)
  at android.content.ContextWrapper.startService (ContextWrapper.java:650)
  at org.altbeacon.beacon.service.Callback.call (Callback.java:60)
  at org.altbeacon.beacon.service.BeaconService.processRangeData (BeaconService.java:360)
  at org.altbeacon.beacon.service.BeaconService.access$300 (BeaconService.java:69)
  at org.altbeacon.beacon.service.BeaconService$1.onCycleEnd (BeaconService.java:320)
  at org.altbeacon.beacon.service.scanner.CycledLeScanner.finishScanCycle (CycledLeScanner.java:241)
  at org.altbeacon.beacon.service.scanner.CycledLeScanner.scheduleScanCycleStop (CycledLeScanner.java:233)
  at org.altbeacon.beacon.service.scanner.CycledLeScanner$1.run (CycledLeScanner.java:229)
  at android.os.Handler.handleCallback (Handler.java:790)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6494)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:438)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)

Regards

dopry commented:
@angel-menendez, can you open a new issue.  with specifics of your setup and we can look into it? A bit of sample code and the error would be appreciated.

goj60 commented:
I did the changes dopry suggests for plugin.xml but still the app crashes...

 Increasing code cache capacity to 512KB
07-12 18:53:03.807 17164-17219/io.ionic.starter E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
    Process: io.ionic.starter, PID: 17164
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:365)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:257)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:784)
     Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/support/v4/content/LocalBroadcastManager;
        at org.altbeacon.beacon.BeaconLocalBroadcastProcessor.unregister(BeaconLocalBroadcastProcessor.java:82)
        at org.altbeacon.beacon.BeaconLocalBroadcastProcessor.register(BeaconLocalBroadcastProcessor.java:74)
        at org.altbeacon.beacon.service.ScanJobScheduler.ensureNotificationProcessorSetup(ScanJobScheduler.java:69)
        at org.altbeacon.beacon.service.ScanJobScheduler.schedule(ScanJobScheduler.java:122)
        at org.altbeacon.beacon.service.ScanJobScheduler.applySettingsToScheduledJob(ScanJobScheduler.java:85)
        at org.altbeacon.beacon.service.ScanJobScheduler.applySettingsToScheduledJob(ScanJobScheduler.java:92)
        at org.altbeacon.beacon.BeaconManager.applyChangesToServices(BeaconManager.java:962)
        at org.altbeacon.beacon.BeaconManager.startRangingBeaconsInRegion(BeaconManager.java:814)
        at com.unarin.cordova.beacon.LocationManager$18.run(LocationManager.java:891)
        at com.unarin.cordova.beacon.LocationManager$33.doInBackground(LocationManager.java:1433)
        at com.unarin.cordova.beacon.LocationManager$33.doInBackground(LocationManager.java:1427)
        at android.os.AsyncTask$2.call(AsyncTask.java:345)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:257) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:784) 
     Caused by: java.lang.ClassNotFoundException: Didn't find class "android.support.v4.content.LocalBroadcastManager" on path: DexPathList[[zip file "/data/app/io.ionic.starter-BOARuijD1wGkOKGEDYAgjg==/base.apk"],nativeLibraryDirectories=[/data/app/io.ionic.starter-BOARuijD1wGkOKGEDYAgjg==/lib/arm64, /data/app/io.ionic.starter-BOARuijD1wGkOKGEDYAgjg==/base.apk!/lib/arm64-v8a, /system/lib64, /vendor/lib64, /product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at org.altbeacon.beacon.BeaconLocalBroadcastProcessor.unregister(BeaconLocalBroadcastProcessor.java:82) 
        at org.altbeacon.beacon.BeaconLocalBroadcastProcessor.register(BeaconLocalBroadcastProcessor.java:74) 
        at org.altbeacon.beacon.service.ScanJobScheduler.ensureNotificationProcessorSetup(ScanJobScheduler.java:69) 
        at org.altbeacon.beacon.service.ScanJobScheduler.schedule(ScanJobScheduler.java:122) 
        at org.altbeacon.beacon.service.ScanJobScheduler.applySettingsToScheduledJob(ScanJobScheduler.java:85) 
        at org.altbeacon.beacon.service.ScanJobScheduler.applySettingsToScheduledJob(ScanJobScheduler.java:92) 
        at org.altbeacon.beacon.BeaconManager.applyChangesToServices(BeaconManager.java:962) 
        at org.altbeacon.beacon.BeaconManager.startRangingBeaconsInRegion(BeaconManager.java:814) 
        at com.unarin.cordova.beacon.LocationManager$18.run(LocationManager.java:891) 
        at com.unarin.cordova.beacon.LocationManager$33.doInBackground(LocationManager.java:1433) 
        at com.unarin.cordova.beacon.LocationManager$33.doInBackground(LocationManager.java:1427) 
        at android.os.AsyncTask$2.call(AsyncTask.java:345) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:257) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:784) 
07-12 18:53:03.815 17164-17164/io.ionic.starter D/CordovaActivity: Paused the activity.
07-12 18:53:03.902 17164-17219/io.ionic.starter I/Process: Sending signal. PID: 17164 SIG: 9

dopry commented:
@goj60, can you open a new issue with specifics about your setup? Cordova version, ionic version, cordova-android version,  plugin version, and Android version?

petermetz commented:
The new release just dropped on npm. Ended up calling it v3.7.0
